### ISSUE TITLE
Similar implemenation of pytree registration via inheritance

### ISSUE THREAD

**Issue by niloch** [NeilGirdhar/cooperative_dataclasses#1]
Hi Neil, 

I saw your comments in the Jax issues on the discussion of dataclasses as pytrees, and some of your desired features.  

I have a toy project I've been noodling around with as a way of learning about jax that I think is close to solving some of these as well.  Here is a link to the base class [Module](https://github.com/niloch/colin_net/blob/67abfba76d6fe41747dbb509d738ed8e37bafbdf/src/colin_net/base.py#L21)

The goal is to let you write code like so:

```python
class Layer(Module, is_abstract=True):
    """Abstract Class for Layers. Enforces subclasses to implement
    __call__"""

    def __call__(self, inputs: Tensor, **kwargs: Any) -> Tensor:
        raise NotImplementedError


class Linear(Layer):
    """Dense Linear Layer.
    Computes output = np.dot(w, inputs) + b"""

    w: Tensor
    b: Tensor

    @jit
    def __call__(self, inputs: Tensor, **kwargs: Any) -> Tensor:
        """outputs = np.dot(w, inputs) + b in single instance notation."""

        return np.dot(self.w, inputs) + self.b

    @classmethod
    def initialize(
        cls,
        *,
        input_size: int,
        output_size: int,
        key: PRNGKey,
    ) -> "Linear":
        """Factory for new Linear from input and output dimentsions"""
        return cls(
            w=random.normal(key, shape=(output_size, input_size)),
            b=np.zeros(shape=(output_size,)),
        )
```
My implementation is a bit messy, but I thought you might be interested.  I'm chasing down some exceptions related to when functions are jit compiled, but grad seems to be working.  Might not work for your usecase completely, but if you have ideas let me know.

**Comment by NeilGirdhar** [comment]
Hey Colin.  Your code looks pretty clean to me!  I think we're trying to solve a similar problem.I do like that you're doing everything using inheritance and `__init_subclass__` rather than decoration and I like your type annotations.

A couple comments:

- You and I both seem to like to write object oriented code, which is fine, but you can't decorate your methods with custom gradients until https://github.com/google/jax/pull/2487 is pulled in, so beware 😄 
- I noticed that you're annotating your metadata with `Any`, but really I think it should be `Hashable` since otherwise you will run into silent trouble when you pass your objects to jitted functions (they will always induce recompilation).  Also, if you ever let a `JaxprTracer` escape as metadata, you have a leaked tracer, which are impossible to debug.  Ensuring that you only output hashable in this codepath prevents both problems.

Here's my code, in case you want to lift any ideas into yours, please feel free:

```python
from numbers import Number
from typing import Any, Hashable, List, Sequence, Tuple, Type, TypeVar

import colorful as cf
import cooperative_dataclasses as dataclasses
import jax
import numpy as np
from jax.interpreters.partial_eval import JaxprTracer
from jax.interpreters.xla import DeviceArray

from .annotations import PyTree

__all__: List[str] = []


T = TypeVar('T', bound=Any)


def dataclass(clz: Type[T]) -> Type[T]:
    """
    some_member: SomeType = field(pytree_like=False)
    """
    # pylint: disable=protected-access
    data_clz: Type[T] = dataclasses.dataclass(frozen=True)(clz)  # type: ignore
    hashed_fields: List[str] = []
    tree_fields: List[str] = []
    for field_info in dataclasses.fields(data_clz):  # type: ignore
        if not field_info.init:
            continue
        if field_info.metadata.get('pytree_like', True):
            tree_fields.append(field_info.name)
        else:
            hashed_fields.append(field_info.name)

    def display(self: T, indent: int) -> str:
        # pylint: disable=protected-access
        retval = cf.violet(type(self).__name__)
        for field_info in dataclasses.fields(data_clz):  # type: ignore
            value = getattr(self, field_info.name)
            if isinstance(value, JaxprTracer):
                value_type = cf.magenta(f"Tracer {value.shape}")
            elif isinstance(value, np.ndarray):
                value_type = cf.yellow(f"NumPy Array {value.shape}")
            elif isinstance(value, DeviceArray):
                value_type = cf.cyan(f"Jax Array {value.shape}")
            elif isinstance(value, Number):
                value_type = cf.orange(value)
            elif hasattr(value, 'display'):
                value_type = value.display(indent + 2)
            else:
                value_type = cf.red(value)
            retval += (" " * (indent + 2)
                       + cf.blue(field_info.name) + cf.base00("=") + value_type + "\n")
        return retval

    def __repr__(self: T) -> str:
        return str(self.display(0))

    def replace(self: T, **updates: Any) -> T:
        """"Returns a new object replacing the specified fields with new values."""
        return dataclasses.replace(self, **updates)

    data_clz.display = display  # type: ignore
    data_clz.__repr__ = __repr__  # type: ignore
    data_clz.__str__ = __repr__  # type: ignore
    data_clz.replace = replace  # type: ignore
    data_clz.hashed_fields = hashed_fields
    data_clz.tree_fields = tree_fields

    def tree_flatten(x: T) -> Tuple[Sequence[PyTree], Hashable]:
        hashed = tuple(getattr(x, name) for name in hashed_fields)
        trees = tuple(getattr(x, name) for name in tree_fields)
        return trees, hashed

    def tree_unflatten(hashed: Hashable, trees: Sequence[PyTree]) -> T:
        if not isinstance(hashed, tuple):
            raise TypeError
        hashed_args = dict(zip(hashed_fields, hashed))
        tree_args = dict(zip(tree_fields, trees))
        return data_clz(**hashed_args, **tree_args)

    jax.tree_util.register_pytree_node(data_clz,
                                       tree_flatten,
                                       tree_unflatten)

    return data_clz


def field(pytree_like: bool = True, **kwargs: Any) -> dataclasses.Field:
    return dataclasses.field(metadata={**kwargs.pop('metadata', {}),
                                       'pytree_like': pytree_like},
                             **kwargs)
```

**Comment by niloch** [comment]
Thanks for the feedback and recommendations!  Sorry for the long gap, [I was finally able to spend some time on this again after a busy period](https://github.com/niloch/colin_net/pull/1).  I have been chasing leaked Tracers, just as you predicted. I'm still not sure if I've handled them correctly the way you recommended. I had a 2 layer MLP that worked fine with jitted functions, but if I increase it to 3 layers (Wat!), it caused a leaked Tracer, possibly due to a for-loop in my predict.  I fixed it by setting `self` as a static argnum on the outer predict functions, but that feels like maybe just a band-aid to an underlying implementation problem. 

I also created `RNGWrapper` to make keys hashable and therefore able to be set as static attributes.   In the `tree_unflatten` method, a new RNGWrapper instance is created from a split
This made it unnecessary to pass keys along with the inputs to all the predict functions.

I've somehow cause my LSTMCell to become unhashable, which is my next bug to nail down.

**Comment by NeilGirdhar** [comment]
> I had a 2 layer MLP that worked fine with jitted functions, but if I increase it to 3 layers (Wat!), it caused a leaked Tracer, 

That's a classic mistake.  Sounds like you're treating an argument as static, and a tracer is getting in.  Write a shim around `jit`, and possibly `grad`, etc. that ensures that static arguments are always hashable.  Also, make sure you don't use `nondiff_argnums` because it is broken for now and it can cause that error.

> I fixed it by setting self as a static argnum on the outer predict functions, but that feels like maybe just a band-aid to an underlying implementation problem.

Yup, bandaid.  I'm assuming `self` is a pytree.

> I also created RNGWrapper to make keys hashable and therefore able to be set as static attributes. In the tree_unflatten method, a new RNGWrapper instance is created from a split
This made it unnecessary to pass keys along with the inputs to all the predict functions.

I did the same thing!  By the way, you don't need to decompose the rng in flatten and unflatten.  Just output it as a pytree.  Here's my version:

```
from __future__ import annotations

from typing import Any, List, Optional, Tuple, TypeVar

import jax.random
from jax import numpy as jnp

from .annotations import Shape, Tensor
from .dataclass import dataclass

__all__ = ['Generator']


T = TypeVar('T', bound='Generator')


@dataclass
class Generator:

    key: Tensor

    def __init__(self,
                 *,
                 seed: Optional[int] = None,
                 key: Optional[Tensor] = None,
                 **kwargs: Any):
        super().__init__(**kwargs)
        if key is None:
            if seed is None:
                raise ValueError
            key = jax.random.PRNGKey(seed)
        object.__setattr__(self, 'key', key)

    # New methods ----------------------------------------------------------------------------------
    def split(self, n: int = 2) -> List[Generator]:
        keys = jax.random.split(self.key, n)
        return [Generator(key=key) for key in keys]

    def normal(self, std_dev: Tensor, shape: Shape = ()) -> (
            Tuple[Generator, Tensor]):
        g1, g2 = self.split()
        return g1, std_dev * jax.random.normal(g2.key, shape)

    def gamma(self, gamma_shape: Tensor, shape: Shape = ()) -> (
            Tuple[Generator, Tensor]):
        g1, g2 = self.split()
        return g1, jax.random.gamma(g2.key, gamma_shape, shape)

    # Magic methods --------------------------------------------------------------------------------
    def __repr__(self) -> str:
        return f"{type(self).__name__}({self.key})"

    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, Generator):
            return NotImplemented
        return jnp.all(self.key == other.key)

    def __ne__(self, other: Any) -> bool:
        if not isinstance(other, Generator):
            return NotImplemented
        return not self.__eq__(other)

    def __hash__(self) -> int:
        return hash((int(self.key[0]), int(self.key[1])))
```

**Comment by sullivancolin** [comment]
I think that if I want to use `jit` that the `tree_flatten` and `tree_unflatten` methods of the Baseclass cannot use  any if statements that depend on `self`.  I think `flax.struct`, and other implementations avoid this problem by using the decorator pattern, and the conditionals are outside the definitions of `tree_flatten`, and `tree_unflatten`.   

I thought about creating explicit groupings of fields for each "Layer" under `params` and `static`, but I found myself writing too many versions of `tree_flatten` and `tree_unflatten` that I have decided to give up on the inheritance of those methods and just force the user to implement them for each new layer type.  Maybe I'll come back to that later if I think of a work around

**Comment by NeilGirdhar** [comment]
> I think that if I want to use jit that the tree_flatten and tree_unflatten methods of the Baseclass cannot use any if statements that depend on self.

Why do you say that?

> too many versions of tree_flatten and tree_unflatten that I have decided to give up on the inheritance of those methods and just force the user to implement them for each new layer type. 

I don't have any tree flatten or unflatten methods.  I just use dataclass.  Why don't you just do that?
